### DISCUSSION TITLE
Logic Backup in Tickets

### DISCUSSION THREAD

**Discussion by slashtangent** [RetroAchievements/RAWeb#1045] [Ideas]
Can we get automatic logic backup created when a ticket is opened?

As in a ticket is opened and the site will automatically backup the logic at the time that the ticket was opened. This will remove the need for devs to backup the logic manually when fixing tickets.

**Comment by meleu** [comment]
In my dreams we would have a commit history of all changes in the achievements code, just like a git repository. While it's currently not viable, I thought in an alternative solution:

Add a `MemAddr` column to the `Ticket`'s table. So when a ticket is opened, the `Ticket.MemAddr` field would be filled with the current content of `Achievements.MemAddr` of the respective achievement.

Before actually working on it I'd like an input from @luchaos if it's a reasonable solution.

**Comment by MagmarFire** [comment]
I think it should be its own table. My view is that the schema would be something like the following:

* `id` - primary key
* `achievementId` - foreign key
* `version` - int, would have a uniqueness constraint with `achievementId`
* `code` - the achievement's code, varchar(65536)
* `createdBy` - user ID, foreign key for users
* `createdOn` - timestamp when the code was updated

Each time an achievement is updated, a new row would, of course, be added to this table, with `version` being 1 greater than the max in the set of all rows sharing the same `achievementId`. As meleu said, we can also add a column to the tickets table--but I disagree that it should be a copy of the code. Instead, I say we make it a foreign key to the new table's `id` column, so that way, we can simply do a table lookup and pull the data that way.

This design would also allow us to implement a rollback system. Devs can go into an achievement's history, select the version number they want to roll back to, and then *poof*, the application will look up the achievement code in the new table and then update the achievement code directly. This would ensure compatibility with RAIntegration while keeping rollback functionality exclusively on the site itself, which may be what we want anyway.

**Comment by luchaos** [comment]
@MagmarFire i like your thinking :) that's on spot.

v2's table schema for reference - that's how it's done there already (polymorphic, so it does that for achievements, leaderboards and rp)
<img width="810" alt="Screenshot 2021-08-04 at 05 57 47" src="https://user-images.githubusercontent.com/1280590/128120339-f25d4b95-5310-4ed4-a971-ae6c9fc71982.png">

with the parentId it goes one step further - a user can have their own unversioned copy of the logic based on a particular trigger's version. this allows to opt-in/out to one's own versions during development. think of it as a git branch that can be switched to. a push to unofficial will always just update that personal version or create a new one if none exists yet. a push to official would merely increment the version.

this allows for diffs between versions, notifying users that are basing their work on a particular version that a new version was pushed, etc. as it's basically a vcs. i thought about doing it with git right away but writing good git bindings would be a lot of work and overkill for this use case.

all of is this not only is great for having backups and a history but also to finally overcome this dreadful issue of not crediting all developers of an achievement properly.

if a versions table is implemented, the code on the resources themselves can be removed because the latest version will always be the active one.

> we make it a foreign key to the new table's id column, so that way, we can simply do a table lookup and pull the data that way

exactly. on tickets, unlocks, leaderboard scores etc only a particular version of a trigger has to be referenced to know which logic the user was using at that time. in addition, a game session reference is stored, too, which can tell which hash was used.

feel free to implement that in v1, yet, it's already done in v2. so a sync command will have to be written to convert v1 versions to v2, causing more work.

imo **adding just a carbon copy of the logic to tickets would actually be the most efficient solution in v1** as it won't have any side-effects for v2 development.

**Comment by MagmarFire** [comment]
I see an issue with the carbon copy idea that would need to be addressed. The `ReportNotes` column is a `varchar(10000)` type, but achievement code can be as many as 65,536 characters. Regardless of what we do, we'll ultimately need to update the database on the off chance that an achievement that has 10,000+ characters of code is ticketed. This is not a light concern, either: I alone have written a handful of achievements that large, courtesy of RATools.

On top of this, there's also the fact that a large glob of code would potentially make the ticket itself less readable, if not completely unwieldy to scroll through in the browser.

I'll weigh between the backup table or just creating a carbon copy in regards to developmental side effects during the transition, but either way, a database change will need to be made, as well as front-end changes to make the code display less unwieldy. I'd like to avoid showing a bunch of stuff at once if possible.

**Comment by luchaos** [comment]
Do we want to tackle this in v1 or not? V2 has versioned triggers for everything - that's not relevant here though.
Carbon copies won't be an issue. Somebody willing to implement it in v1 is our usual problem.
